### PR TITLE
Helicone Node Package To Support OpenAI v4

### PR DESCRIPTION
**Pull request by colegottdank**
Todo:
- Add tests
- Update docs

### CODE REVIEW

File: helicone-node/UsageExample.ts [hunk 1]
```diff
@@ -1,82 +1,227 @@
...
-
-async function feedback(openAi: HeliconeProxyOpenAIApi) {
-  // Create chat completion
-  const result = await openAi.createChatCompletion({
-    model: "gpt-3.5-turbo",
-    messages: [
-      {
-        role: "user",
-        content: "Say hi!",
+// let feedbackCount = 0;
```
Comment (chitalian) [inline helicone-node/UsageExample.ts:9]: just delete this no?

File: helicone-node/UsageExample.ts [hunk 2]
```diff
@@ -1,82 +1,227 @@
...
+  // console.log(data2.choices[0].text);
+
+  // const { data: data3, response: response3 } = await openai.embeddings
+  //   .create({
+  //     input: "Hello, world!",
+  //     model: "text-embedding-ada-002",
+  //   })
+  //   .withResponse();
+
+  // console.log(data3.data[0].embedding); // Ugly, but works
```
Comment (chitalian) [inline helicone-node/UsageExample.ts:188]: and this

File: helicone-node/UsageExample.ts
Comment (chitalian) [inline helicone-node/UsageExample.ts:1]: We should get rid of this file and just move everything to tests/

File: helicone-node/async_logger/HeliconeAsyncLogger.ts [hunk 3]
```diff
@@ -43,73 +43,65 @@ export enum Provider {
...
     } else if (provider == Provider.AZURE_OPENAI) {
-      options.url = `${basePath}/oai/v1/log`;
+      url = `${basePath}/oai/v1/log`;
     } else if (provider == Provider.ANTHROPIC) {
-      options.url = `${basePath}/anthropic/v1/log`;
+      url = `${basePath}/anthropic/v1/log`;
     } else {
       console.error("Failed to log to Helicone: Provider not supported");
       return;
     }
```
Comment (chitalian) [inline helicone-node/async_logger/HeliconeAsyncLogger.ts:76]: nit: this should be a function
```typescript
function getHeliconeUrl(provider: string, basePath: string): string {
...
}
```

File: helicone-node/async_logger/HeliconeAsyncOpenAI.ts [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+  private originalCompletionCreate: hcone.CompletionCreateFunction;
+  private originalChatCreate: hcone.ChatCreateFunction;
+  private originalEmbeddingCreate: hcone.EmbeddingCreateFunction;
+  private heliconeHeaders: { [key: string]: string };
+  private logger: HeliconeAsyncLogger;
+
+  constructor(private options: IHeliconeAsyncClientOptions) {
+    const {
+      apiKey = Core.readEnv("OPENAI_API_KEY"),
+      organization = Core.readEnv("OPENAI_ORG_ID") ?? null,
```
Comment (chitalian) [inline helicone-node/async_logger/HeliconeAsyncOpenAI.ts:31]: do we always have to pull this out? Is it possible to inherit this behavior from somewhere?

File: helicone-node/async_logger/HeliconeAsyncOpenAI.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    };
+
+    this.helicone = new Helicone(heliconeMeta);
+    this.logger = new HeliconeAsyncLogger(options);
+    this.heliconeHeaders = new HeliconeHeaderBuilder(heliconeMeta)
+      .withPropertiesHeader()
+      .withUserHeader()
+      .build();
+
+    this.monkeyPatch();
```
Comment (chitalian) [inline helicone-node/async_logger/HeliconeAsyncOpenAI.ts:50]: sick

File: helicone-node/async_logger/HeliconeAsyncOpenAI.ts [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+      (loggingStream) => {
+        resolveLoggingStream(loggingStream);
+      }
+    );
+
+    result.finally(() => {
+      this.logRequest(startTime, result, body, isStream, loggingStreamPromise);
+    });
+
+    return result as hcone.CompletionCreateReturnType<T>;
```
Comment (chitalian) [inline helicone-node/async_logger/HeliconeAsyncOpenAI.ts:84]: I didnt know you could namespace types like this. this is dope

File: helicone-node/async_logger/HeliconeLogger.ts [hunk 7]
```diff
@@ -121,18 +118,18 @@ export class HeliconeLogBuilder {
...
+    this.options = options;
+    this.helicone = new Helicone(options);
+    this.logger = new HeliconeAsyncLogger(options);
   }
 
-  submit(request: HeliconeLogBuilder): Promise<AxiosResponse<any, any>> {
+  async submit(request: HeliconeLogBuilder): Promise<Response> {
     const asyncLogRequest = request.build();
-    return this.logger.log(asyncLogRequest, Provider.CUSTOM_MODEL);
+    return await this.logger.log(asyncLogRequest, Provider.CUSTOM_MODEL);
```
Comment (chitalian) [inline helicone-node/async_logger/HeliconeLogger.ts:133]: why does this need to await now?

File: helicone-node/async_logger/HeliconeOpenAITypes.ts
Comment (chitalian) [inline helicone-node/async_logger/HeliconeOpenAITypes.ts:1]: Dope

File: helicone-node/core/HeliconeClientOptions.ts
Comment (chitalian) [inline helicone-node/core/HeliconeClientOptions.ts:1]: Dope

File: helicone-node/proxy_logger/HeliconeProxyOpenAI.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+import { IHeliconeProxyClientOptions } from "../core/HeliconeClientOptions";
+
+export class HeliconeProxyOpenAI extends OpenAI {
+  public helicone: Helicone;
+  private heliconeHeaders: { [key: string]: string };
+
+  constructor(private options: IHeliconeProxyClientOptions) {
+    const {
+      apiKey = Core.readEnv("OPENAI_API_KEY"),
+      organization = Core.readEnv("OPENAI_ORG_ID") ?? null,
```
Comment (chitalian) [inline helicone-node/proxy_logger/HeliconeProxyOpenAI.ts:14]: can we inherit this behavior from somewhere else?

File: helicone-node/tests/async.test.ts
Comment (chitalian) [inline helicone-node/tests/async.test.ts:1]: THIS IS DOPE

Comment (chitalian) [reply helicone-node/tests/async.test.ts:1]: Wait - this isn't being ran?

File: helicone-node/tsconfig.json [hunk 9]
```diff
@@ -1,8 +1,9 @@
 {
   "compilerOptions": {
     "declaration": true,
-    "target": "es6",
-    "module": "commonjs",
+    "target": "es2020",
+    "module": "CommonJS",
+    "moduleResolution": "Node",
```
Comment (chitalian) [inline helicone-node/tsconfig.json:6]: what does this do? Does this match OpenAI's pacakge?

File: tests/python_integration_tests.py [hunk 10]
```diff
@@ -1,7 +1,6 @@
 # TODO deprecate this file and move all test to helicone_python/tests
 
 import requests
-import pytest
```
Comment (chitalian) [inline tests/python_integration_tests.py:4]: alright

File: worker/src/lib/HeliconeHeaders.ts [hunk 11]
```diff
@@ -104,7 +104,7 @@ export class HeliconeHeaders implements IHeliconeHeaders {
       openaiBaseUrl: this.headers.get("Helicone-OpenAI-Api-Base") ?? null,
       retryHeaders: this.getRetryHeaders(),
       promptFormat: this.headers.get("Helicone-Prompt-Format") ?? null,
-      requestId: this.headers.get("Helicone-Request-Id") ?? null,
+      requestId: this.headers.get("Helicone-Request-Id") ?? crypto.randomUUID(),
```
Comment (chitalian) [inline worker/src/lib/HeliconeHeaders.ts:107]: nice catch. I am pretty sure we do this somewhere else - nice to do it earlier thought

**Review by chitalian — APPROVED** [review]

File: helicone-node/tsconfig.json (hunk 9 above)
Comment (colegottdank) [reply helicone-node/tsconfig.json]: matching OpenAI's package: https://github.com/openai/openai-node/blob/master/tsconfig.json#L4C3-L7C19

File: helicone-node/tests/async.test.ts
Comment (colegottdank) [reply helicone-node/tests/async.test.ts:1]: It was, for some reason it stopped, taking a look

File: helicone-node/async_logger/HeliconeLogger.ts (hunk 7 above)
Comment (colegottdank) [reply helicone-node/async_logger/HeliconeLogger.ts:133]: I think it should have always been awaited. The user can just decide to not await, but having it here gives them the option

File: helicone-node/proxy_logger/HeliconeProxyOpenAI.ts (hunk 8 above)
Comment (colegottdank) [reply helicone-node/proxy_logger/HeliconeProxyOpenAI.ts]: it is actually not needed as the super does that

File: helicone-node/tests/async.test.ts
Comment (colegottdank) [reply helicone-node/tests/async.test.ts:1]: It is now, not sure what changed

**Pull request merged**